pkg/controller/ipsec/ipsec_controller.go

```diff
@@ -0,0 +1,185 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const ipsecSignerName = "network.openshift.io/ipsec"
+
+var ca *crypto.TLSCertificateConfig
+
+// Add and Start it when the Manager is Started.
+func Add(mgr manager.Manager, status *statusmanager.StatusManager) error {
+	return add(mgr, newReconciler(mgr, status))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager) reconcile.Reconciler {
+	return &ReconcileCSR{client: mgr.GetClient(), scheme: mgr.GetScheme(), status: status}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("ipsec-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CetificateSigningRequest resource
+	err = c.Watch(&source.Kind{Type: &csrv1.CertificateSigningRequest{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Make an ephemeral (in-memory) CA.
+	// TODO: This is temporary as it should use the Operator PKI ovn-ca secret
+	ca, _ = crypto.MakeSelfSignedCAConfig("ipsec-ca", 365)
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileCSR{}
+
+// ReconcileCSR reconciles a cluster CertificateSigningRequest object
+type ReconcileCSR struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	status *statusmanager.StatusManager
+}
+
+// Reconcile
+func (r *ReconcileCSR) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	csr := &csrv1.CertificateSigningRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, csr)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Println("Object seems to have been deleted")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Println(err)
+
+		return reconcile.Result{}, err
+	}
+
+	// Only handle CSRs for this signer
+	if csr.Spec.SignerName != ipsecSignerName {
+		return reconcile.Result{}, nil
+	}
+
+	cert, err := extractCertificateRequest(csr.Spec.Request)
+	if err != nil {
+		log.Println("Could extract certificate from request")
+		return reconcile.Result{}, nil
+	}
+	cacert, err := extractCert(ca)
+	if err != nil {
+		log.Println("Could not extract CA certificate")
+		return reconcile.Result{}, nil
+	}
+
+	signedCert, err := signCertificate(newCertificateTemplate(), cert.PublicKey, cacert, ca.Key)
+	if err != nil {
+		log.Println("Could not sign request")
+		return reconcile.Result{}, nil
```

Comment:
err?, same for the other places, seems on err we always return nil

Reply:
In some cases, I don't think I should return error as the reconciliation will be requeued - especially in cases in which the CSR is malformed. In those cases, I should set the status appropriately and not try to reconcile again. TBH, I havent though too much about this yet as I just wanted to get the basics working.

Reply:
Yup, your instinct is correct - return `nil`, set status.